File: samples/sample-java/src/main/java/sample/SampleApp.java
package sample;

import kiit.codes.Checked;
import kiit.codes.Checks;
import kiit.codes.CodesToHttp;
import kiit.codes.Err;
import kiit.codes.Failed;
import kiit.codes.Passed;
import kiit.codes.Status;
import kiit.codes.StatusException;
import kiit.codes.StatusExceptions;
import kiit.codes.formats.Catalog;
import kiit.codes.formats.CodeDetail;
import kiit.codes.formats.CodeDetails;
import kiit.codes.formats.CodesToProblem;
import kiit.codes.formats.ErrorDetail;
import kiit.codes.formats.ErrorItem;
import kiit.codes.formats.Problem;

import java.util.List;
import java.util.Map;

/**
 * Living documentation of kiit-codes from plain Java. Each section below only compiles because
 * of the @JvmStatic/@JvmField/@JvmOverloads/@file:JvmName annotations added to the library, and
 * the pattern-matching switch requires the sealed hierarchies to emit PermittedSubclasses
 * (jvmTarget = JVM_21). See kiit-codes/src/jvmTest/java/kiit/codes/JavaInteropTest.java for the
 * same surface exercised as assertions.
 */
public class SampleApp {

    public static void main(String[] args) {
        // @JvmField companion constants: Passed.Succeeded.SUCCESS, not .Companion.getSUCCESS()
        Status ok = Passed.Succeeded.SUCCESS;
        Status denied = Failed.Restricted.DENIED;
        System.out.println("ok status: " + describe(ok));
        System.out.println("denied status: " + describe(denied));

        // @JvmStatic + @JvmOverloads: Err.of(message), no explicit null for the trailing Throwable
        Err err = Err.of("email is required");
        System.out.println("err: " + err.getMessage());

        // @JvmOverloads on the primary constructor: no-arg CodesToHttp()
        CodesToHttp http = new CodesToHttp();
        System.out.println("http code for ok: " + http.toCode(ok));
        System.out.println("http code for denied: " + http.toCode(denied));

        // @JvmStatic + @JvmOverloads: Checked.success(), no explicit Passed argument
        Checked validEmail = Checked.success();
        Checked invalidEmail = Checked.failure(Failed.Invalid.BAD_REQUEST, List.of(err));
        System.out.println("validEmail.isValid = " + validEmail.isValid());
        System.out.println("invalidEmail.isValid = " + invalidEmail.isValid());

        // @file:JvmName("Checks"): kiit.codes.Checks, not CheckedKt
        Checked combined = Checks.collect(validEmail, invalidEmail);
        System.out.println("combined.isValid = " + combined.isValid() + ", errors = " + combined.getErrors().size());

        // @JvmOverloads on the StatusException subclasses: only the required status argument
        try {
            throw new StatusException.RestrictedException(Failed.Restricted.UNAUTHENTICATED);
        } catch (StatusException e) {
            System.out.println("caught: " + e.getStatus().getName() + " — " + e.getMessage());
        }

        // @file:JvmName("StatusExceptions") + @JvmOverloads: no explicit errors list needed
        StatusException fromStatus = StatusExceptions.toException(Failed.Invalid.NOT_FOUND);
        System.out.println("converted: " + fromStatus.getClass().getSimpleName());

        // Custom, consumer-defined Status: a non-kiit origin plus an internal-organization scope.
        // Failed.Rejected has no @JvmOverloads, so every constructor parameter must be supplied.
        Failed.Rejected duplicateCharge =
                new Failed.Rejected(
                        "DUPLICATE_CHARGE",
                        "This charge has already been processed",
                        "com.stripe",
                        "payments.cards");
        System.out.println("http code: " + http.toCode(duplicateCharge));

        // Catalog supplies baseUrl per origin; a built-in Status needs no registration, it
        // defaults to kiit-codes' own taxonomy docs.
        Catalog catalog = Catalog.of(Map.of("com.stripe", "https://stripe.com/problems"));
        CodesToProblem problems = new CodesToProblem(catalog, http);

        // Two independent converters off the same Status, pick whichever fits the boundary:

        // CodesToProblem.build (@JvmOverloads): the RFC 9457 shape, for an HTTP API response.
        Problem<ErrorDetail> stripeProblem = problems.build(duplicateCharge);
        System.out.println("[rfc]  type: " + stripeProblem.getType());
        System.out.println("[rfc]  title: " + stripeProblem.getTitle());
        System.out.println("[rfc]  status: " + stripeProblem.getStatus());

        // @file:JvmName("CodeDetails") + @JvmOverloads: kiit-codes' own shape, no baseUrl or HTTP
        // status needed. Useful for internal service-to-service calls and background jobs.
        CodeDetail<ErrorDetail> stripeCode = CodeDetails.toCodeDetail(duplicateCharge);
        System.out.println("[kiit] path: " + stripeCode.getPath());
        System.out.println("[kiit] code: " + stripeCode.getCode());
        System.out.println("[kiit] success: " + stripeCode.getSuccess());
        System.out.println("[kiit] message: " + stripeCode.getMessage());

        Problem<ErrorDetail> kiitProblem = problems.build(Failed.Invalid.NOT_FOUND);
        System.out.println("kiit problem type: " + kiitProblem.getType());

        // Custom error shape: supply your own ErrorItem when field + message isn't enough.
        Err.ErrorList validationErr =
                new Err.ErrorList(List.of(Err.on("phone", "1234567890123", "Too long")), "Validation failed");
        CodeDetail<DetailedError> customCode =
                CodeDetails.toCodeDetail(
                        Failed.Invalid.INVALID_VALUE,
                        validationErr,
                        e ->
                                new DetailedError(
                                        e instanceof Err.ErrorField f ? f.getField() : null,
                                        e.getMessage(),
                                        "check formatting"));
        System.out.println("[custom] code: " + customCode.getCode());
        System.out.println("[custom] errors: " + customCode.getErrors());
    }

    // Custom ErrorItem: field + message plus a hint, for when the default ErrorDetail shape
    // (field + message only) isn't enough.
    static final class DetailedError implements ErrorItem {
        private final String field;
        private final String message;
        private final String hint;

        DetailedError(String field, String message, String hint) {
            this.field = field;
            this.message = message;
            this.hint = hint;
        }

        @Override
        public String getField() {
            return field;
        }

        @Override
        public String getMessage() {
            return message;
        }

        @Override
        public String toString() {
            return "DetailedError(field=" + field + ", message=" + message + ", hint=" + hint + ")";
        }
    }

    // JDK 21 pattern-matching switch, exhaustive with no `default` branch. Only compiles because
    // Status/Passed/Failed are sealed and Kotlin emitted PermittedSubclasses at jvmTarget 21.
    private static String describe(Status status) {
        return switch (status) {
            case Passed.Succeeded s -> "Succeeded: " + s.getName();
            case Passed.Pending p -> "Pending: " + p.getName();
            case Passed.Excluded e -> "Excluded: " + e.getName();
            case Passed.Information i -> "Information: " + i.getName();
            case Failed.Restricted r -> "Restricted: " + r.getName();
            case Failed.Invalid i -> "Invalid: " + i.getName();
            case Failed.Rejected r -> "Rejected: " + r.getName();
            case Failed.Unserved u -> "Unserved: " + u.getName();
        };
    }
}

File: samples/sample-kotlin/src/main/kotlin/sample/SampleApp.kt
package sample

import kiit.codes.Checked
import kiit.codes.CodesToHttp
import kiit.codes.Err
import kiit.codes.Excluded
import kiit.codes.Failed
import kiit.codes.Information
import kiit.codes.Invalid
import kiit.codes.Passed
import kiit.codes.Pending
import kiit.codes.Rejected
import kiit.codes.Restricted
import kiit.codes.Status
import kiit.codes.StatusException
import kiit.codes.Succeeded
import kiit.codes.Unserved
import kiit.codes.code
import kiit.codes.formats.Catalog
import kiit.codes.formats.CodesToProblem
import kiit.codes.formats.ErrorItem
import kiit.codes.formats.toCodeDetail
import kiit.codes.path
import kotlin.random.Random

private val http = CodesToHttp()
private val catalog = Catalog.of(mapOf("com.stripe" to "https://stripe.com/problems"))
private val problems = CodesToProblem(catalog, http)

fun main() {
    test0()
    test1()
    test2()
    test3()
    test4()
}

fun test1() {
    val service = UserService()

    report("create alice", service.create("alice", "alice@example.com"))
    report("create alice again", service.create("alice", "alice@example.com"))
    report("create with blank email", service.create("bob", ""))

    report("authorize alice as alice", service.authorize("alice", "alice"))
    report("authorize alice as bob", service.authorize("alice", "bob"))
    report("authorize unknown user", service.authorize("carol", "carol"))

    try {
        service.requireAuthorized("alice", "bob")
    } catch (e: StatusException) {
        println("caught StatusException: ${e.status.name} — ${e.message}")
    }
}

private fun report(label: String, status: Status) {
    val outcome =
        when (status) {
            is Passed -> "ok"
            is Failed -> "failed"
        }
    println("$label -> ${status.name} ($outcome, http=${http.toCode(status)})")
}

fun test0() {
    val status =
        Failed.Invalid(
            name = "CREATED",
            message = "failure",
            origin = "kiit"
        )

    println(status.success) // false
    println(CodesToHttp().toCode(status)) // 400
}

fun test2() {
    val value = Random.nextInt(1, 8)
    val result: Status = when (value) {
        1 -> Succeeded.SUCCESS
        2 -> Pending.ACCEPTED
        3 -> Excluded.SKIPPED
        4 -> Information.NOTICE
        5 -> Restricted.DENIED
        6 -> Invalid.INVALID_VALUE
        7 -> Rejected.RULE_VIOLATION
        else -> Unserved.UNEXPECTED
    }

    val name = when(result) {
        is Passed -> {
            when(result) {
                is Passed.Succeeded   -> "Succeeded"
                is Passed.Pending     -> "Pending"
                is Passed.Excluded    -> "Excluded"
                is Passed.Information -> "Information"
            }
        }
        is Failed -> {
            when(result) {
                is Failed.Restricted  -> "Restricted"
                is Failed.Invalid     -> "Invalid"
                is Failed.Rejected    -> "Rejected"
                is Failed.Unserved    -> "Unserved"
            }
        }
    }
    println("${name.padEnd(11)}: ${result.name}")
}


fun test3() {
    val check1 = validatePhone("")
    val check2 = validatePhone("12345678901")
    val check3 = validatePhone("1111111111")
    val check4 = validatePhone("1234567890")
    val check5 = validatePhone("9876543210")
    val check6 = validatePhone("1234567890", "admin")
    println(check1)
    println(check2)
    println(check3)
    println(check4)
    println(check5)
    println(check6)
}

fun test4() {
    // Custom, consumer-defined Status: a non-kiit origin plus an internal-organization scope,
    // distinct from kiit-codes' own built-in registry (see Codes.kt).
    val duplicateCharge =
        Rejected(
            name = "DUPLICATE_CHARGE",
            message = "This charge has already been processed",
            origin = "com.stripe",
            scope = "payments.cards",
        )
    println("path: ${duplicateCharge.path}") // com.stripe:payments.cards
    println("code: ${duplicateCharge.code}") // Failed:Rejected:DUPLICATE_CHARGE

    // Two independent converters off the same Status, pick whichever fits the boundary:

    // CodesToProblem.build: the RFC 9457 shape, for an HTTP API response. baseUrl comes from
    // catalog, registered above for "com.stripe" (a built-in Status needs no registration, it
    // defaults to kiit-codes' own https://kiit.dev/problems).
    val stripeProblem = problems.build(duplicateCharge)
    println("[rfc]  type: ${stripeProblem.type}")
    println("[rfc]  title: ${stripeProblem.title}")
    println("[rfc]  status: ${stripeProblem.status}")

    // toCodeDetail: kiit-codes' own shape, no baseUrl or HTTP status needed, since path/code are
    // already a complete identity. Useful for internal service-to-service calls, background jobs,
    // and anywhere else an HTTP-shaped response doesn't apply.
    val stripeCode = toCodeDetail(duplicateCharge)
    println("[kiit] path: ${stripeCode.path}")
    println("[kiit] code: ${stripeCode.code}")
    println("[kiit] success: ${stripeCode.success}")
    println("[kiit] message: ${stripeCode.message}")

    // An Err.ErrorList populates errors[], one ErrorDetail per wrapped Err, in both shapes.
    val validationErr =
        Err.ErrorList(
            errors = listOf(Err.on("phone", "1234567890123", "Too long")),
            message = "Validation failed",
        )
    val validationProblem = problems.build(Invalid.INVALID_VALUE, validationErr)
    val validationCode = toCodeDetail(Invalid.INVALID_VALUE, validationErr)
    println("[rfc]  validation type: ${validationProblem.type}")
    println("[kiit] validation code: ${validationCode.code}")
    println("[kiit] validation errors: ${validationCode.errors}")

    // Custom error shape: supply your own ErrorItem when field + message isn't enough.
    data class DetailedError(override val field: String?, override val message: String, val hint: String) : ErrorItem
    val customCode =
        toCodeDetail(Invalid.INVALID_VALUE, validationErr) { err ->
            DetailedError((err as? Err.ErrorField)?.field, err.message, hint = "check formatting")
        }
    println("[custom] code: ${customCode.code}")
    println("[custom] errors: ${customCode.errors}")
}

fun validatePhone(phone: String, caller: String = "guest"): Checked {
    return when {
        phone.isEmpty()    -> {
            Checked.failure(
                Invalid.INVALID_VALUE,
                listOf(Err.on("phone", phone, "Too short")),
            )
        }
        phone.length > 10 -> {
            Checked.failure(
                Invalid.INVALID_VALUE,
                listOf(Err.on("phone", phone, "Too long")),
            )
        }
        phone == "1111111111" -> {
            Checked.failure(
                Rejected.RULE_VIOLATION,
                listOf(Err.on("phone", phone, "Reserved phone for testing")),
            )
        }
        phone.startsWith("123") && caller != "admin" -> {
            Checked.failure(
                Restricted.DENIED,
                listOf(Err.on("phone", phone, "Only admins can validate internal-use numbers")),
            )
        }
        else -> {
            Checked.success()
        }
    }
}
